# PYTHON

Lenguaje de programación no compilado. Se usa:

- Cuando es importante poder hacer cosas de manera interactiva
- Cuando no importa la velocidad del código

- Mucho mas sencillo que lenguajes tradicionales. 
- Es un lenguaje de muy alto nivel, muy poderoso
- Filosofı́a ”baterı́as incluidas”: viene con muchas librerı́as para facilitar tareas de diversos
tipos

1. Información: http://www.python.org
2. Tutorial básico: http://docs.python.org/tut/

Interprete se llama python3 (versión mas avanzada) o ipython3 (i = interactivo)

## Instalación de python

$ sudo apt-get install python3 python3-pip

$ python3 -m pip install numpy scipy matplotlib jupyter timetable pandas


## Jupyter

__Jupyter__ es un interprete de __python__ (entre otros lenguajes de programación) que ofrece una shell interactiva vía web, a la que podemos acceder desde un navegador

$ jupyter notebook

o pueden utilizar __Jupyter__ sin instalar nada en [Binder](http://jupyter.org/try) o en [Colaboratory](https://colab.research.google.com/)

# Evaluar una expresión matemática 

Consideren la expresión de la altura de una pelota que es lanzada en dirección vertical
$$
y(t) = y_0 + v_0~t - \frac{1}{2} ~g~t^2
$$
donde
* $y$ es la altura (posición) como función del tiempo
* $y_0$ es la posición inicial en $t=0$
* $v_0$ es la velocidad inicial en $t=0$
* $g$ es la aceleración de la gravedad

**Determina la posición $y$, dado que conoces $y_0 = 0$, $v_0 = 5$ m/s, $g=9.81$ m/s$^2$ y $t=0.6$ s.**

In [1]:
print (0 + 5*0.6 - 1/2*9.81*0.6**2)

1.2342


* En esta expresión se usan los cuatro operadores estándares de aritmética $+, -, * , /$
* Además la potencia es con doble asterisco: $(0.6)^2$ es 0.6**2
* **Python** procede de izquierda a derecha, término a término (con términos separados por $+$ o $-$). En cada término las operaciones de potencia, tienen precedencia sobre la multiplicación y la división
* Puedes usar paréntesis para controlar la manera en la que se evalua una expresión.

# Usando variables

**Ahora evalúa la posición $y$ para diferentes $t$ usando variables.**

In [2]:
v0 = 5
g = 9.81
t = 0.6
y = v0*t - 0.5*g*t**2
print (y)

1.2342


* Las variables en **Python** se definen 

**nombre = valor o expresión con variables que ya se definieron**

* Usa nombres de variables similares a las de expresiones matemáticas. Los nombres de variables pueden tener minúsculas y mayúsculas (**Python** las distingue!), números 0 al 9 y guión bajo, pero el primer caracter no puede ser número

*  Esta versión es más práctica, porque puedes cambiar el valor de $t$ en un solo lugar al evaluar

Nombres de variable reservados: *and , as , assert , break , class , continue , def , del , elif , else , except , False ,
finally , for , from , global , if , import , in , is , lambda , None , nonlocal ,
not , or , pass , raise , return , True , try , with , while , yield*

Si quieres usar un nombre reservado, puedes agregar un guión bajo al final, por ejemplo *lambda_*

La función *dir()* en python es una función que devuelve una lista de los atributos y métodos de cualquier objeto(funciones, módulos, cadenas, listas, diccionarios, etc.)

In [3]:
import math
dir(math)

['__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'lcm',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'nextafter',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc',
 'ulp']

In [4]:
# Programa para calcular la altura de una pelota en movimiento vertical
v0 = 5 #velocidad inicial
g = 9.81 #aceleracion de la gravedad
t = 0.6 #tiempo 
y = v0*t - 0.5*g*t**2 #expresion de la cinematica
print (y)

1.2342


# Documenta tu código

La documentación de tu código será crucial para que puedas colaborar y extender las capacidades al integrar/merger varios códigos. En este ejemplo la documentación es sencilla, pero siempre tienes que documentar lo mejor posible!

In [5]:
# Programa para calcular la altura de una pelota en movimiento vertical
v0 = 5 #velocidad inicial
g = 9.81 #aceleracion de la gravedad
t = 0.6 #tiempo 
y = v0*t - 0.5*g*t**2 #expresion de la cinematica

# Formato de texto y números 

Ahora imprime el valor de $y$ de manera que el programa nos reporte: **En $t=0.6$ s, la altura de la pelota es $1.23$ m.** También queremos tener control de la precisión del resultado. Por ejemplo en este problema tenemos precisión de cm. En **Python** se usa la estructura de sintaxis de https://en.wikipedia.org/wiki/Printf_format_string, heredada de programación en C.

In [6]:
# una manera de imprimir el resultado es
print('En t=%g s, la altura de la pelota es %.2f m.' % (t, y))
# otra manera de reportar el resultado es
print("En t={0} s, la altura de la pelota es {1} m." .format(t, y))
#una ultima
print(f'En t={t} s, la altura de la pelota es {y} m.')

En t=0.6 s, la altura de la pelota es 1.23 m.
En t=0.6 s, la altura de la pelota es 1.2342 m.
En t=0.6 s, la altura de la pelota es 1.2342 m.


# Ejercicio 1

En tu portafolio de clase, elabora un programa que calcule la temperatura en grados Celsius a partir de temperaturas conocidas (dadas) en grados Farenheit.

# Funciones matemáticas básicas

Determina el tiempo $t$, que le toma a la pelota, alcanzar una altura dada $y_c$ dado que conoces $y_0 = 0$, $v_0 = 5$ m/s y $g=9.81$ m/s$^2$.} Sabemos que
$$
y(t) = v_0~t - \frac{1}{2} ~g~t^2.
$$
Entonces si $y(t)$ ahora es conocida y dada por $y_c$, queremos resolver
$$
\frac{1}{2} ~g~t^2 - v_0~t + y_c =0,
$$
cuyas soluciones son
$$
t_1 = \frac{v_0 - \sqrt{v_0^2 - 2~g~y_c}}{g} ~~~~~~~~~ t_2 = \frac{v_0 + \sqrt{v_0^2 - 2~g~y_c}}{g}
$$
Las soluciones son el tiempo $t=t_1$ cuando la pelota va hacia arriba, y el tiempo $t=t_2 > t_1$, cuando la pelota va hacia abajo.

En **Python** las funciones matemáticas básicas, tales como la raíz cuadrada, las funciones trigonométricas, los logaritmos, etc., estan definidas en el módulo https://docs.python.org/3/library/math.html. Para usarlas, necesitas cargar el módulo y para llamarlas usas la sintaxis: **math.nombre()**

In [7]:
v0 = 5
g = 9.81
yc = 0.2
import math
t1 = (v0 - math.sqrt(v0**2 - 2*g*yc))/g
t2 = (v0 + math.sqrt(v0**2 - 2*g*yc))/g
print ('En t=%g s y %g s, la altura de la pelota es %g m.' % (t1,t2,yc))

En t=0.0417064 s y 0.977662 s, la altura de la pelota es 0.2 m.


Para no arrastrar el modulo cada vez que necesitas una función, puedes cargar la función desde el principio asi: 

**from math import sqrt**

o puedes cargar todas las funciones del módulo asi: 

**from math import $*$**



# Ejercicio 2


Considera la definición del $\sinh(x)$
$$
\sinh(x) = \frac{e^x - e^{-x}}{2}.
$$

En tu portafolio de clase, elabora un programa que calcule el valor de $\sinh(x)$ en $x=2\pi$ de tres diferentes maneras:
1. evaluando $\sinh(x)$ directamente
2. evaluando con la definición del lado derecho, usando la función exponencial
3. evaluando con la definición del lado derecho, usando la potencia

# Error de redondeo

En el ejercicio anterior, imprime los resultados con 16 decimales

In [8]:
print ('%.16f %.16f' % (1/49.0*49, 1/51.0*51))

0.9999999999999999 1.0000000000000000


Puedes ver que dos resultados coinciden pero el tercero no. Porque?

Los cálculos se hacen con números reales, que en principio requieren un número infinito de decimales para representarlos. Entonces, en los cálculos se truncan los decimales y se genera un error por redondeo. Típicamente se usan 17 dígitos para un número real. Por ejemplo, calcula $1/49*49$ y por otro lado, calcula $1/51*51$. Ambos deberían ser 1, pero en un caso el error se propaga al resultado y en el otro no. 

# Números complejos

En **Python** la unidad de los imaginarios es **1j**, en lugar de la **i** que se utiliza en las matemáticas. Puedes definir un complejo con la notación **a + b*1j** o con la notación **complex(a,b)**.

In [9]:
u = 5.3 + 2.1j #define un número complejo
v = 3 #define un entero
w = u + v #sumalos
w #checa el resultado

(8.3+2.1j)

In [10]:
a = -1
b = 0.8
s = a + b*1j #define otro complejo
s 

(-1+0.8j)

In [11]:
a = -1
b = 0.8
s = complex(a, b) #define otro complejo
s

(-1+0.8j)

In [12]:
u = 5.3 + 2.1j
v = 3
w = u + v
a = -1
b = 0.8
s = a + b*1j #define otro complejo
s*w #multiploca complejo*complejo

(-9.98+4.540000000000001j)

In [13]:
u = 5.3 + 2.1j
v = 3
w = u + v
a = -1
b = 0.8
s = a + b*1j
s/w #divide complejo/complejo

(-0.09031377899045019+0.11923601637107775j)

Un objeto llamado **w** del tipo **complex**, tiene la funcionalidad de poder extraer la parte real o imaginaria y de poder calcular el complejo conjugado.

In [14]:
a = 4
b = -2.3
w = a + b*1j
w.real

4.0

In [15]:
a = 4
b = -2.3
w = a + b*1j
w.imag

-2.3

In [16]:
a = 4
b = -2.3
w = a + b*1j
w.conjugate()

(4+2.3j)

# Funciones de variables complejas

Para evaluar funciones básicas de matemáticas con valores complejos, no debes usar el módulo **math**, porque ese módulo asume que la evaluación es con números reales (flotante). Para evaluar funciones matemáticas básicas en variable compleja, usa el módulo https://docs.python.org/3/library/cmath.html

In [17]:
from cmath import sin, sinh
r1 = sin(8j)
r1

1490.4788257895502j

In [18]:
from cmath import sin, sinh
r2 = 1j*sinh(8)
r2

1490.4788257895502j

# Ejercico 3

1. Considera la relación entre el seno en variable compleja y el seno hiperbólico en variable real $x$,
$$
\sin(ix) = i\sinh(x).
$$

   En tu portafolio de clase, elabora un programa que calcule el valor de $\sin(ix)$ y de $\sinh(x)$ para ciertos valores dados de $x$, para verificar la identidad.

2. Considera la relación de Euler para $x$ real,
$$
e^{ix} = \cos(x) + i\sin(x).
$$

En tu portafolio de clase, elabora un programa que calcule el valor de $\cos(x), \sin(x)$ y de $e^{ix}$ para ciertos valores dados de $x$, para verificar la identidad

# Funciones en el plano complejo

En lugar de cargar ciertas funciones con **math** y **cmath**, puedes cargar el módulo **Numpy** de la librería **Scimath**, o de plano cargar la librería completa **Scitools**. Cualquiera de las siguientes declaraciones funciona
* from numpy.lib.scimath import *
* from scipy import *
* from scitools.std import *

para carga funciones que pueden ser evaluadas en el plano complejo (que incluye los reales), las funciones corren un poco más lento que las de **math, cmath**, pero ya no te preocupas del dominio de evaluación.

In [19]:
from numpy.lib.scimath import *
sqrt(4)

2.0

In [20]:
from numpy.lib.scimath import *
sqrt(-4)

2j

# Ejercicio 4

Este tratamiento flexible de funciones en el plano complejo permite encontrar las raices reales o complejas, de una función cuadrática. Considera que las raices de $f(z)=az^2+bz+c$ se obtienen
$$
z^{\pm} = \frac{-b \pm \sqrt{b^2-4ac}}{2a}.
$$

En tu portafolio de clase, elabora un programa en el que uses **Numpy** para calcular el valor de las raices con diferentes valores dados de $a,b,c$, para obtener ejemplos de raices reales y complejas.

# Derivadas e integrales simbólicas

Podemos ver como funciona la diferenciación e integración simbólica, usando la expresión $v_0 t - 1/2 g t^2$ 



In [21]:
from sympy import (
symbols, # define simbolos para matematicas simbolicas
diff,    # diferenciacion simbolica de expresiones
integrate, # integracion simbolica de expresiones
Rational, # define numeros racionales
lambdify, # convierte expresiones simbolicas en funciones de Python
)
t, v0, g = symbols('t v0 g')
y = v0*t - Rational(1,2)*g*t**2
print ('Primera derivada:', diff(y,t))
print ('Segunda derivada:', diff(y, t, t))
deriv=diff(y,t)


Primera derivada: -g*t + v0
Segunda derivada: -g


**NOTA**: Definimos a **t, v0, g** como símbolos, no como flotantes. Por lo tanto, todas las definiciones **y, deriv** son tambien expresiones simbólicas (no numéricas).

Podemos regresar a la expresión $v_0 t - 1/2 g t^2$ numérica, usando **lambdify**

In [22]:
print ('Regresamos a la original', integrate(deriv,t))
v = lambdify([t, v0, g], # argumentos en v
deriv)  # expresion simbolica que estamos convirtiendo en numerica
print ('evaluacion', v(2, 5, 9.81))

Regresamos a la original -g*t**2/2 + t*v0
evaluacion -14.620000000000001


# Resolución de ecuaciones

Una ecuación lineal, definida a partir de una expresión **e** que es cero, puede ser resuelta con **solve(e,t)**, donde **t** es la incógnita (definida como símbolo) en la ecuación. Por ejemplo, podemos encontrar las raices de **y = 0**

In [23]:
from sympy import (
symbols, # define simbolos para matematicas simbolicas
Rational, # define numeros racionales 
solve # resuelve ecuaciones
)
t, v0, g = symbols('t v0 g') 
y = v0*t - Rational(1,2)*g*t**2
roots = solve(y, t) 
print ('Las raices son', roots) 
print ('La primera raiz es', roots[0])
print ('La primera raiz es', roots[1])


Las raices son [0, 2*v0/g]
La primera raiz es 0
La primera raiz es 2*v0/g


Y puedes checar las respuestas insertando las raices en **y = 0**

In [24]:
print ('Check de primera raiz',y.subs(t,roots[0]))
print ('Check de primera raiz',y.subs(t,roots[1]))

Check de primera raiz 0
Check de primera raiz 0


# Serie de Taylor

La representación de la expresión *expr* en Serie de Taylor de orden **n**, en la variable **t** alrededor de **t0**, se calcula como **expr.series(t, t0, n)**. Por ejemplo, podemos encontrar la representación en Serie de Taylor de $e^t$

In [25]:
from sympy import symbols, exp, sin, cos
t = symbols('t')
f = exp(t)
f.series(t, 0, 3)


1 + t + t**2/2 + O(t**3)

In [26]:
from sympy import symbols, exp, sin, cos
t = symbols('t')
f = exp(sin(t))
f.series(t, 0, 8)

1 + t + t**2/2 - t**4/8 - t**5/15 - t**6/240 + t**7/90 + O(t**8)

Puedes manipular las expresiones con **expand** y **simplify**

In [27]:
from sympy import simplify, expand
x, y = symbols('x y')
f = -sin(x)*sin(y) + cos(x)*cos(y)
simplify(f)

cos(x + y)

In [28]:
from sympy import simplify, expand
x, y = symbols('x y')
expand(sin(x+y), trig=True) # con un hint

sin(x)*cos(y) + sin(y)*cos(x)

Puedes exportar las respuestas en formato **LaTex**

In [29]:
from sympy import latex, symbols, exp, sin, cos
t = symbols('t')
print (latex(f.series(t, 0, 7)))

- \sin{\left(x \right)} \sin{\left(y \right)} + \cos{\left(x \right)} \cos{\left(y \right)}


# En resumen

**VARIABLES:** El enunciado

        alguna_variable = obj

define una variable con el nombre **alguna_variable** que esta referida al objeto **obj**.

* Los nombres de variables pueden tener núeros, guión bajo, letras mayúsculas y minúsculas pero no pueden empezar con un número. 

* Usa nombres que estén conectados con la interpretación de esa variable en el problema. 

* Buenos nombres de variables, reducen la necesidad de comentar el programa. Todo lo que se encuentre después del **#** es ignorado.

**OBJETOS:** En **Python** hay diferentes tipos de objetos. Hasta ahora hemos trabajado con:

       integer -- números enteros, objecto tipo **int**  

       float -- números con decimales, objeto tipo **float**

       string -- pedazo de texto, objeto tipo **str**  



In [30]:
a = 'Este es un pedazo de texto \n escrito en dos lineas.'
b = "Los textos pueden escribirse con una comilla o doble comilla" 
c = """Los textos con tres dobles comillas 
pueden ocupar 
varias lineas.
"""
print (c)

Los textos con tres dobles comillas 
pueden ocupar 
varias lineas.



**OPERADORES:**  

Los operadores en expresiones artiméticas siguen las reglas de las matemáticas: la potencia es evaluada antes de la multiplicación y división y éstas últimas se evaluan antes de la suma y la resta.

Estas reglas se reescriben, si se usan paréntesis. Se sugiere el uso de paréntesis de cualquier manera, para mejorar la claridad en el cédigo.

Con cuidado cuando programamos divisiones! los denominadores con operaciones de suma y resta requieren paréntesis. No es lo mismo **a/(a+b)** que **(a/a+b)**


**FUNCIONES MATEMÁTICAS**

El módulo **math** contiene las funciones matemáticas más usadas para variable real. Los módulos se importan antes de usarse. Puedes importar módulos de diferentes maneras:

In [31]:
# Importar modulo - las functiones requiren prefijo 
import math 
a = math.sin(math.pi*1.5)

# Importar funciones individuales - no se requiere prefijo
from math import sin, pi
a = sin(pi*1.5)

# Importar todo el modulo - no se requiere prefijo \\
from math import *
a = sin(pi*1.5)

# Ejercicio 5

¿Cuál es la trayectoria de una pelota que se lanza con una rapidéz inicial $v_0$ a un ángulo $\theta$ medido de la horizontal?

Sabemos que la pelota seguirá una trayectoria $y = f(x)$, donde, en ausencia de resistencia del aire,

$$
f(x) = x \tan \theta - \frac{g}{2v_0^2}\frac{x^2}{\cos^2\theta} + y_0.
$$

En esta expresión, $x$ es la coordenada horizontal, $g$ es la aceleración de la gravedad y $y_0$ es la posición inicial de la pelota.

1. En tu portafolio de clase, elabora un programa en el que evalues esta expresión. El programa debe escribir el valor de todas las variables involucradas junto con las unidades usadas.